File: Golang/MinHeap.go
package MinHeap

import "errors"

type MinHeap struct {
	arr []int
}

func (this *MinHeap) Insert(val int) {
	this.arr = append(this.arr, val)
	this.bubbleUp()
	return
}

func (this *MinHeap) Pop() int {
	pop := this.arr[0]
	this.arr[len(this.arr)-1], this.arr[0] = this.arr[0], this.arr[len(this.arr)-1]
	this.arr = this.arr[:len(this.arr)-1]
	this.bubbleDown()
	return pop
}

func (this *MinHeap) Peek() (error, int) {
	if len(this.arr) > 0 {
		return nil, this.arr[0]
	} else {
		return errors.New("There is no number in heap"), 0
	}

}

func (this *MinHeap) bubbleDown() {
	current := 0
	for current <= len(this.arr)-1 {
		leftChild := current*2 + 1
		if leftChild > len(this.arr)-1 {
			return
		}
		rightChild := current*2 + 2
		if rightChild > len(this.arr)-1 {
			if this.arr[current] > this.arr[leftChild] {
				this.arr[current], this.arr[leftChild] = this.arr[leftChild], this.arr[current]
				current = leftChild
				continue
			}
			return
		}
		compare := 0
		if this.arr[leftChild] < this.arr[rightChild] {
			compare = leftChild
		} else {
			compare = rightChild
		}
		if this.arr[current] > this.arr[compare] {
			this.arr[current], this.arr[compare] = this.arr[compare], this.arr[current]
			current = compare
		} else {
			return
		}
	}
}

func (this *MinHeap) bubbleUp() {
	current := len(this.arr) - 1
	for current > 0 {
		parent := (current - 1) / 2
		if this.arr[current] < this.arr[parent] {
			this.arr[current], this.arr[parent] = this.arr[parent], this.arr[current]
			current = parent
		} else {
			return
		}
	}
}

File: Typescript/MinHeap.ts
type HeapNode = number

class MinHeap {
    private heap: HeapNode[];
    constructor() {
        this.heap = [];
    }

	get length(): number{
		return this.heap.length
	}
	
    public peek() {
		if(this.length > 0)
			return this.heap[0];
		else 
			return undefined
    }

    public push(val: HeapNode) {
        this.heap.push(val);
        this.bubbleUp();
    }
    
    public pop() {
        if(this.length === 0) {
            return undefined
        } 
        [this.heap[0], this.heap[this.length - 1]] = [this.heap[this.length - 1],this.heap[0]]
        const pop = this.heap.pop();
        this.bubbleDown();
        return pop;
    }

    private bubbleUp() {
        let current = this.length - 1;
        while (current > 0) {
            let parent = Math.ceil(current / 2) - 1;
            if (this.compare(this.heap[current], this.heap[parent]) == -1) {
				[this.heap[current], this.heap[parent]] = [ this.heap[parent], this.heap[current]];
                current = parent;
            } else {
                return;
            }
        }
    }

    private bubbleDown() {
        let current = 0;
        while (true) {
            let left = current * 2 + 1;
            if (left >= this.length) return;
            let right = left + 1;
            if (right >= this.length) right = left;
            let smaller: number;
            if (this.compare(this.heap[left], this.heap[right]) == -1) {
                smaller = left;
            } else {
                smaller = right;
            }
            if (this.compare(this.heap[smaller], this.heap[current]) == -1) {
                [this.heap[smaller], this.heap[current]] = [
                    this.heap[current],
                    this.heap[smaller],
                ];
                current = smaller;
            } else {
                return;
            }
        }
    }

    private compare(node1: HeapNode, node2: HeapNode): number {
        if (node1 > node2) {
            return 1;
        }
        if (node1 == node2) {
            return 0 
        }
        return -1;
    }
}